core/server/session_logs.go

```diff
@@ -114,35 +126,191 @@ func (cs *coreServer) GetSessionLogs(ctx context.Context, msg *pb.GetSessionLogs
 		},
 	)
 	if err != nil {
-		return nil, err
+		return &pb.GetSessionLogsResponse{Error: err.Error()}, err
+	}
+
+	var (
+		logEntries []*pb.LogEntry
+		firstToken string
+	)
+
+	if msg.GetLogSourceFilter() == "" || msg.GetLogSourceFilter() == logger.SessionLogSource {
+		// get gitops-run logs
+		gitopsRunLogs, token, err := getGitOpsRunLogs(
+			ctx,
+			sessionName,
+			msg.GetToken(),
+			asS3Reader(minioClient),
+			logger.SessionLogBucketName,
+			msg.GetLogLevelFilter())
+		if err != nil {
+			return &pb.GetSessionLogsResponse{Error: err.Error()}, err
+		}
+		logEntries = append(logEntries, gitopsRunLogs...)
+
+		firstToken = token
 	}
 
-	logs, lastToken, err := getLogs(ctx, msg.GetSessionId(), msg.GetToken(), asS3Reader(minioClient), logger.SessionLogBucketName)
+	// get pod-logs
+	podLogs, secondToken, logSources, err := getPodLogs(
+		ctx,
+		msg.GetToken(),
+		asS3Reader(minioClient),
+		logger.PodLogBucketName,
+		msg.GetLogSourceFilter(),
+		msg.GetLogLevelFilter())
 	if err != nil {
-		return nil, err
+		return &pb.GetSessionLogsResponse{Error: err.Error()}, err
 	}
+	logEntries = append(logEntries, podLogs...)
 
-	logEntries := []*pb.LogEntry{}
+	// we sort the logs by timestamp
+	sort.Slice(logEntries, func(i, j int) bool {
+		tsi, _ := time.Parse(time.RFC3339, logEntries[i].GetTimestamp())
+		tsj, _ := time.Parse(time.RFC3339, logEntries[j].GetTimestamp())
+		return tsi.Before(tsj)
+	})
 
-	for _, log := range logs {
-		logEntry := &pb.LogEntry{}
-		err = json.Unmarshal([]byte(log), &logEntry)
+	return &pb.GetSessionLogsResponse{
+		Logs:       logEntries,
+		NextToken:  firstToken + "," + secondToken,
+		LogSources: append([]string{logger.SessionLogSource}, logSources...),
+	}, nil
+}
+
+func detectLogLevel(message string) string {
```

Comment:
❔ In regular pod logs, obtained with `kubectl logs`, I see a log `level` field:

```
{"level":"info","ts":"2023-02-08T22:43:02.558Z","caller":"podinfo/main.go:151","msg":"Starting podinfo","version":"6.3.3","revision":"ae3fe3da989f5c0173ad9477975cca185b280455","port":"9898"}
{"level":"info","ts":"2023-02-08T22:43:02.559Z","caller":"api/server.go:223","msg":"Starting HTTP Server.","addr":":9898"}
```

Is something like that present in logs, collected by Fluent-Bit, and can we use it in principle instead of this custom function? Or do we need to use our own log levels which do not match default pod log levels?

Reply:
YES!

I would have detected it with this before falling back to the detectLogLevel function.